pennylane/gradients/hadamard_gradient.py

```diff
@@ -0,0 +1,318 @@
+
+        gradient_data.append(num_tape)
+
+    def processing_fn(results):  # pylint: disable=too-many-branches
+        # TODO interface compatibility
+        multi_measurements = len(tape.measurements) > 1
+        multi_params = len(tape.trainable_params) > 1
+
+        grads = []
+        final_res = []
+        for coeff, res in zip(coeffs, results):
+            if isinstance(res, tuple):
+                final_res.append([2 * coeff * r for r in res])
+            else:
+                final_res.append(2 * coeff * res)
+
+        # Post process for probs
+        if measurements_probs:
+            projector = np.array([1, -1])
+            for idx, res in enumerate(final_res):
+                if multi_measurements:
+                    for prob_idx in measurements_probs:
+                        prob = tape.measurements[prob_idx]
+                        wires = prob.wires
+                        num_wires = len(wires)
+                        res_reshaped = qml.math.reshape(res[prob_idx], (2**num_wires, 2))
+                        final_res[idx][prob_idx] = qml.math.tensordot(
+                            res_reshaped, projector, axes=[[1], [0]]
+                        )
+                else:
+                    prob_idx = measurements_probs[0]
+                    prob = tape.measurements[prob_idx]
+                    wires = prob.wires
+                    num_wires = len(wires)
+                    res = qml.math.reshape(res, (2**num_wires, 2))
+                    final_res[idx] = qml.math.tensordot(res, projector, axes=[[1], [0]])
+
+        for idx, num_tape in enumerate(gradient_data):
+            if num_tape is None:
+                grads.append(qml.math.zeros(()))
+            elif num_tape == 1:
+                grads.append(final_res[idx])
+            else:
+                grads.append(sum(final_res[idx : idx + num_tape]))
+
+        if not multi_measurements and not multi_params:
+            return np.array(grads[0])
+
+        if (multi_params and not multi_measurements) or (not multi_params and multi_measurements):
+            grads_tuple = tuple(grads)
+            return grads_tuple
+
+        # Reordering to match the right shape for multiple measurements
+        grads_reorder = [[0] * len(tape.trainable_params) for _ in range(len(tape.measurements))]
+
+        for i in range(len(tape.measurements)):
+            for j in range(len(tape.trainable_params)):
+                grads_reorder[i][j] = grads[j][i]
+
+        # To tuple
+        if multi_params:
+            grads_tuple = tuple(tuple(elem) for elem in grads_reorder)
+        else:
+            grads_tuple = tuple(elem[0] for elem in grads_reorder)
+
+        return grads_tuple
+
+    return g_tapes, processing_fn
+
+
+def _get_generators(trainable_op):
+    """From a trainable operation, extract the unitary generators and their coefficients."""
+    # For PhaseShift, we need to separate the generator in two unitaries (Hardware compatibility)
```

Comment:
I will double check that, thank you for raising this point!

Reply:
Tell me if I am wrong but I think it can add a term:

$\bra{o}\bra{\psi} PauliY \otimes Hamiltonian \ket{0}\ket{\psi}=\bra{0} PauliY \ket{0} + \bra{\psi} Hamiltonian \ket{\psi}= 0 +  \bra{\psi} Hamiltonian \ket{\psi}$

What do you think?

Reply:
Ah, no, the tensor product decomposes as product, not as sum, right?
So I think we would get
$\bra{0}\bra{\psi} PauliY \otimes Hamiltonian \ket{0}\ket{\psi}=\bra{0} PauliY \ket{0} \times \bra{\psi} Hamiltonian \ket{\psi}= 0 \times  \bra{\psi} Hamiltonian \ket{\psi}=0$

Reply:
Thanks, all clear!